File: 0332_Reconstruct_Itinerary.java
// DFS and Backtracking
class Solution {
    public List<String> findItinerary(List<List<String>> tickets) {
        tickets.sort((t1, t2) -> { // sort tickets by to-airport, lexographically
            return t1.get(1).compareTo(t2.get(1));
        });
            
        ArrayList<String> order = new ArrayList<>(); // order of itinerary
        boolean[] used = new boolean[tickets.size()];
        
        HashMap<String, List<Integer>> connections = new HashMap<>(); // (airport) : {ticket ids that are from airport}
        
        // populate connections with directed edges
        for(int i = 0; i < tickets.size(); i++) {
            String from = tickets.get(i).get(0);
            if(connections.get(from) == null) connections.put(from, new ArrayList<>());
            connections.get(from).add(i);
        }
        
        dfs(tickets, order, used, "JFK", connections); // we assume that this returns true and correctly fills order
        return order;
    }
    
    boolean dfs(List<List<String>> tickets, List<String> order, boolean[] used, String current, 
                     HashMap<String, List<Integer>> connections) {
        if(order.size() == used.length) {
            order.add(current);
            return true;
        }
        
        if(connections.get(current) == null) return false;
        for(Integer ticket : connections.get(current)) { // since tickets is sorted, this will greedily choose
                                                         // the lowest ticket, lexographically
            if(!used[ticket]) {
                used[ticket] = true;
                order.add(tickets.get(ticket).get(0));
                if(dfs(tickets, order, used, tickets.get(ticket).get(1), connections)) return true;
                order.remove(order.size() - 1);
                used[ticket] = false;
            }
        }
        
        return false;
    }
}

// Eulerian Path
class Solution {
    public List<String> findItinerary(List<List<String>> tickets) {
        HashMap<String, PriorityQueue<String>> outEdges = new HashMap<>();
        // use dfs to find an Eulerian path; use priority queues to explore edges in lexographical order
        
        for(List<String> ticket : tickets) { // populate outEdges
            String from = ticket.get(0);
            String to = ticket.get(1);
            
            outEdges.putIfAbsent(from, new PriorityQueue<>());
            outEdges.get(from).add(to);
        }
        
        LinkedList<String> solution = new LinkedList<>();
        findEulerianPath("JFK", outEdges, solution);
        return solution;
    }
    
    private void findEulerianPath(String current, HashMap<String, PriorityQueue<String>> outEdges,
                                  LinkedList<String> result) {
        
        // depth-first explore outgoing edges until the node "gets stuck"
        if(outEdges.get(current) != null) {
            while(!outEdges.get(current).isEmpty()) {
                findEulerianPath(outEdges.get(current).poll(), outEdges, result);
            }
        }
        
        // when the node is "stuck", it can be appended to the front of the solution
        result.add(0, current);
    }
}

File: 0332_Reconstruct_Itinerary.py
from typing import List
from collections import defaultdict

class Solution:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        # greedy graph traversal

        # use adjacency dict instead of adjacency list to get ordered multiset
        cnx = defaultdict(lambda: defaultdict(int)) # [airport_from] -> ([airport_to] -> count)

        for (frum, to) in sorted(tickets):
            cnx[frum][to] += 1

        def min_lex_traverse(start: str) -> List[str]:
            curr = start
            res = [start]

            while cnx[curr]:
                to = next(iter(cnx[curr]))
                cnx[curr][to] -= 1

                if cnx[curr][to] == 0:
                    del cnx[curr][to]

                res.append(to)
                curr = to

            return res

        res = min_lex_traverse("JFK")

        # now we have reached a dead-end (we know res[0] is the correct start and res[-1] must be the correct
        # end node because they both have indegree != outdegree)
        # it is possible, however, that not all edges are being used
        # but since we know that there exists a path that uses all edges, this implies that
        # the graph formed by the remaining edges must only contain nodes of matching indegree and outdegree
        # (w/r/t the remaining edges)
        # => all remaining nodes are part of cycles (which can be traversed greedily w/o the concern of reaching a dead-end)
        # we must add these cycles to `res`, however, by definition, `res` is already at a lexicographical minimum
        # and we expicitly made each choice in `res` accordingly, therefore all new cycles should
        # begrudglingly added as far back in `res` as possible.
        # using the same greedy algorithm

        def rec(old):
            # assemble `old` backwards, populating cycles
            old2_reversed = []

            for airport in reversed(old):
                cycle = min_lex_traverse(airport)
                if len(cycle) > 1:
                    cycle2 = rec(cycle)
                    old2_reversed += reversed(cycle2)
                else:
                    old2_reversed += reversed(cycle)

            return list(reversed(old2_reversed))

        return rec(res)

        # this solution is slop, but I'll leave it
        # I realize it probably could be implemented with a single recursive dfs
        # but it took a while to get this far
        # this one's pretty tricky from scratch because it's both:
        # 1. Eulerian Path Finding
        # 2. Realization that lexicographical reverse-assembly will work
        # considerably easier if you have either or both of those fresh in your mind, but
        # when cold, it's a bit wicked